resources/session03/calculator.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+
+# These would usually be in a templates dir someplace.
+# I <3 jinja
+
+
+layout = """\
+<html>
+<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css" integrity="sha384-1q8mTJOASx8j1Au+a5WDVnPi2lkFfwwEAa8hDDdjZlpLegxhjVME1fgjWPGmkzs7" crossorigin="anonymous">
+<head>
+<title>Overly Complex Calculator</title>
+</head>
+<body>
+<div class="col-md=4"></div>
+<div class="jumbotron vertical-center col-md-4">
+<div class="container-fluid text-center">
+{input}
+</div>
+</div>
+<div class="col-md=4"></div>
+</body>
+</html>"""
+
+root = """\
+<form role="form">
+  <div class="form-group">
+    <label for "a">First value:</label>
+    <input type="number" class="form-control" name="a">
+  </div>
+  <div class="form-group">
+    <label for "b">Second value:</label>
+    <input type="number" class="form-control" name="b">
+  </div>
+  <button type="submit" formaction="add" class="btn btn-default">Add</button>
+  <button type="submit" formaction="subtract" class="btn btn-default">Subtract</button>
+  <button type="submit" formaction="multiply" class="btn btn-default">Multiply</button>
+  <button type="submit" formaction="divide" class="btn btn-default">Divide</button>
+</form>
+"""
+
+my_template = """\
+<h3>{a} {verb} {b} equals {c}</h3>
+<a class="btn btn-primary" href="/" role="button">Again?</a>
+"""
+
+error_template = """\
+<h3>OMG SOMETHING HAPPENED</h3>
+<p>{error}</p>
+<a class="btn btn-primary" href="/" role="button">Go Home</a>
+"""
+
+
+# Here lies the application section.  Normally would be a module someplace else
+def render(template, **kwargs):
+    '''In a real app, this would be built to allow inheritance
+       between templates.  It's late, and I'm cheating.'''
+    template = layout.format(input=template)
+    return template.format(**kwargs)
+
+
+class Erlenmeyer:
+    '''Sorry for the pun'''
+    # dict for mapping urls to views
+    paths = {}
+
+    def add_path(self, path, f):
+        self.paths[path] = f
+
+    def route(self, path):
+        '''Decorator to add view functions.  Way more complex than needed
+           for this trivial case, but this made it interesting :-p'''
+        def decorated(f):
+            self.add_path(path, f)
+            return f
+        return decorated
+
+    def _parse_qs(self, query):
+        '''Return dict built from query string parameters'''
+        d = {}
+        if '&' in query:  # multiple params
+            for item in query.split('&'):
+                a, b = item.split('=')
+                d.update({a: b})
```

Comment:
This will only ever allow a single value for any given query parameter.  Granted that is all you need for this particular app, but how might you go about solving the problem in a more generalized fashion?